File: 2021/day-02/day02-2.js
const fs = require('fs');

let horizontal_position = 0;
let depth = 0;
let aim = 0;
const datas = fs.readFileSync('input.txt', 'utf-8').split(/\n/).map(line => {
    let results = line.split(' ');
    return { direction: results[0], value: parseInt(results[1]) };
});
datas.map( data => {
    switch(data['direction']) {
        case 'forward':
            horizontal_position += data['value'];
            depth += aim * data['value'];
            break;

        case 'up':
            aim -= data['value'];
            break;

        case 'down':
            aim += data['value'];
            break;
    }
});
console.log(horizontal_position * depth);

File: 2021/day-02/day02-2.cpp
#include <iostream>
#include <fstream>

class Point
{
public:
    Point()
        : x(0)
        , y(0)
        , aim(0)
    {}

    void Forward(int offset)
    {
        x += offset;
        y += aim * offset;
    }
    void Up(int offset) { aim -= offset; }
    void Down(int offset) { aim += offset; }
    
    int GetResult() const { return x * y; }

private:
    int x;
    int y;

    int aim;

};

int main(int argc, char** argv)
{
    Point position;
    std::ifstream file("input.txt");
    std::string direction;
    std::string value;
    while (file >> direction, file >> value)
    {
        auto v = std::stoi(value);
        if (direction == "forward")
        {
            position.Forward(v);
        }
        else if (direction == "up")
        {
            position.Up(v);
        }
        else if (direction == "down")
        {
            position.Down(v);
        }
    }

    std::cout << position.GetResult();

    return 0;
}
